### PR TITLE
explicit FE mechanics solver

### PR DESCRIPTION
**Pull request by boyceg**
### IBAMR Pull Request Checklist
- [x] Does the test suite pass?
- [x] Was clang-format run?
- [x] Were relevant issues cited? Please remember to add `Fixes #12345` to close
      the issue automatically if we are fixing the problem.
- [x] Is this a change others will want to know about? If so, then has a
      changelog entry been added?
- [x] If new data structures have been added to a class then have they been set
      up appropriately for restarts? If so, ensure that the restart version
      number is incremented.
- [x] Does this change include a bug fix or new functionality? If so, a new test
      or tests should be added. New tests should run quickly (less than a minute
      in release mode). If possible, an older test should gain a new option so
      that we do not need to compile more test executables.
- [x] Did you (if your account has permission to do so) set relevant labels on
      GitHub for the pull request?

### CODE REVIEW

**Comment by drwells** [comment]
:tada:

I'll take a close look later today.

**Comment by boyceg** [comment]
This does not yet have stabilization (to reduce locking) or generalized-alpha (to improve stability in time stepping). I thought that maybe those should go in separate PRs.

**Comment by boyceg** [comment]
I just noticed a few issues in the comments, now fixed.

**Comment by boyceg** [comment]
@benlvr, could you add simplified versions of the various solid mechanics benchmarks we've used with IBFE as examples in a separate PR?

File: src/IB/FEMechanicsExplicitIntegrator.cpp [hunk 1]
```diff
@@ -0,0 +1,739 @@
...
+        IBTK_CHKERRQ(ierr);
+        ierr = VecWAXPY(d_X_vecs->get("new", part).vec(),
+                        dt,
+                        d_U_vecs->get("current", part).vec(),
+                        d_X_vecs->get("current", part).vec());
+        IBTK_CHKERRQ(ierr);
+    }
+    // We copy F^{n} into F^{n+1} to make sure that it is stored in the viz
+    // file.
+    d_F_vecs->copy("current", { "new" });
```
Comment (drwells) [inline src/IB/FEMechanicsExplicitIntegrator.cpp:180]: Will this actually save output? I believe the only vectors libMesh saves are the solutions.

Comment (boyceg) [reply src/IB/FEMechanicsExplicitIntegrator.cpp:180]: The end-time step routine copies new into current and solution.

Comment (drwells) [reply src/IB/FEMechanicsExplicitIntegrator.cpp:180]: :+1:

File: src/IB/FEMechanicsExplicitIntegrator.cpp [hunk 2]
```diff
@@ -0,0 +1,739 @@
...
+                        d_F_vecs->get("new", part).vec(),
+                        d_U_vecs->get("current", part).vec());
+        IBTK_CHKERRQ(ierr);
+        ierr = VecWAXPY(d_X_vecs->get("new", part).vec(),
+                        dt,
+                        d_U_vecs->get("new", part).vec(),
+                        d_X_vecs->get("current", part).vec());
+        IBTK_CHKERRQ(ierr);
+    }
+}
```
Comment (drwells) [inline src/IB/FEMechanicsExplicitIntegrator.cpp:230]: Could we move these time step routines to the base class? The only difference between these and the ones in IBFEMethod is the addition of the direct forcing kinematics.

File: src/IB/FEMechanicsExplicitIntegrator.cpp [hunk 3]
```diff
@@ -0,0 +1,739 @@
...
+            {
+                F_system.add_variable("F_" + std::to_string(d), d_fe_order_force[part], d_fe_family_force[part]);
+            }
+            F_system.get_dof_map()._dof_coupling = &d_diagonal_system_coupling;
+        }
+
+        setup_system_vectors(
+            &equation_systems, { COORDS_SYSTEM_NAME, VELOCITY_SYSTEM_NAME }, { "current", "half", "new" });
+        setup_system_vectors(
+            &equation_systems, { FORCE_SYSTEM_NAME }, { "current", "half", "new", "tmp", "RHS Vector" });
```
Comment (drwells) [inline src/IB/FEMechanicsExplicitIntegrator.cpp:493]: This is the same code as in IBFEMethod. Could move this also to the base class and then both classes can use that (and IBFEMethod can then set up FEDataManager too)?

File: src/IB/FEMechanicsExplicitIntegrator.cpp [hunk 4]
```diff
@@ -0,0 +1,739 @@
...
+                        if (node->n_dofs(U_sys_num))
+                        {
+                            const int U_dof_index = node->dof_number(U_sys_num, d, 0);
+                            DofConstraintRow U_constraint_row;
+                            U_constraint_row[U_dof_index] = 1.0;
+                            U_dof_map.add_constraint_row(U_dof_index, U_constraint_row, 0.0, false);
+                        }
+                    }
+                }
+            }
```
Comment (drwells) [inline src/IB/FEMechanicsExplicitIntegrator.cpp:593]: Same - lets not duplicate stuff if we can avoid it.

File: src/IB/FEMechanicsExplicitIntegrator.cpp (hunk 2 above)
Comment (boyceg) [reply src/IB/FEMechanicsExplicitIntegrator.cpp:230]: No --- `IBFEMethod` does not time step `U`.

File: src/IB/FEMechanicsExplicitIntegrator.cpp (hunk 3 above)
Comment (boyceg) [reply src/IB/FEMechanicsExplicitIntegrator.cpp:493]: I am not sure that they will ultimately use the same collections of vectors.

Comment (drwells) [reply src/IB/FEMechanicsExplicitIntegrator.cpp:493]: It wouldn't be too hard to have a class set up whatever extra vectors it needs but at the moment they look like the same sets to me (`X`, `dX`, `U`, `F`).

**Comment by drwells** [comment]
@boyceg what is the state of this? If its close we can open issues for the minor cleanup/refactoring and merge so that interested parties can use this.

**Comment by boyceg** [comment]
I agree with the suggestions on factoring out common code but haven't had time to work on it. (I have been focused on trying to stabilize the nearly incompressible case, since we want to try to use it in applications.) It is a little more than just moving common code from the derived classes to the base class, because the two derived classes do slightly different things. I am hoping to revisit later this week. However, I have some time sensitive stuff to take care of this week, so probably this week will slip to next week.

A couple other things that I was thinking about either addressing now or later.

1. Time stepping is handled in a different way here than in most other places in the library. Should it be more uniform?
2. I think we want to make a more general way to set up the systems. The specific use case that I have in mind is to hook together an IBFE-based model to a FE mechanics model. Currently this is harder than it should be because the same `System` names will be used for, e.g., deformation in both models, which I suspect will cause problems for libMesh.

**Comment by boyceg** [comment]
(Item 2 above is my punishment for using static variables to keep track of system names. :-D)

**Comment by drwells** [comment]
We should try to finish this up soon. I'll take another look next week.

**Comment by drwells** [comment]
One thing lead to another and we still haven't done this. @boyceg We could delay the 0.8 release by two weeks if we are sure we can wrap this class up this month - what do you think?

**Comment by boyceg** [comment]
If you could give me this week to try to get it done, I will do my best.

All of my kids will be at home _all the time_ for the following two weeks, and so if it drags beyond this week, I think early January is most likely.

**Comment by drwells** [comment]
I don't want to push the release back too far (I like this twice a year cycle) but I think as long as this is done by January 10 or so we should be on track to put out the release before the semester starts.

**Comment by boyceg** [comment]
@drwells I have factored some common code into `FEMechanicsBase`. Can you take another look?

File: src/IB/FEMechanicsExplicitIntegrator.cpp [hunk 5]
```diff
@@ -0,0 +1,727 @@
...
+    }
+}
+
+void
+FEMechanicsExplicitIntegrator::doInitializeFESystemVectors()
+{
+    FEMechanicsBase::doInitializeFESystemVectors();
+    std::vector<EquationSystems*> equation_systems;
+    for (const auto& es : d_equation_systems) equation_systems.push_back(es.get());
+    d_X_vecs.reset(new LibMeshSystemVectors(equation_systems, COORDS_SYSTEM_NAME));
```
Comment (boyceg) [inline src/IB/FEMechanicsExplicitIntegrator.cpp:533]: @drwells: I cannot think of a simple way to avoid having `FEMechanicsExplicitIntegrator` and `IBFEMethod` both implement this, with the only difference being `LibMeshSystemVectors` vs. `LibMeshIBSystemVectors`.

Comment (boyceg) [reply src/IB/FEMechanicsExplicitIntegrator.cpp:533]: (I _can_ think of complicated ways to do this. 😀 )

Comment (boyceg) [reply src/IB/FEMechanicsExplicitIntegrator.cpp:533]: (For instance using lambda expressions.)

**Comment by boyceg** [comment]
@drwells, @rossisimone, I think this is cleaned up. Can you guys take a look? A lot of code is now factored into `FEMechanicsBase`. I may need to update test outputs if logging outputs wind up changing.

**Comment by boyceg** [comment]
I just rebased and force-pushed to fix the conflict.

**Comment by drwells** [comment]
Nice! I'll dig in tomorrow morning.

**Comment by boyceg** [comment]
OK, current tests pass! I need to add new ones for the mechanics solver itself.

File: ibtk/src/lagrangian/FEDataManager.cpp [hunk 6]
```diff
@@ -2778,9 +2778,19 @@ namespace
...
+        db = input_db->getDatabase("FEProjector");
+    }
+    else
+    {
+        db = new InputDatabase("FEProjector");
+        if (input_db->keyExists("num_fischer_vectors"))
+        {
+            db->putInteger("num_fischer_vectors", input_db->getInteger("num_fischer_vectors"));
+        }
+    }
```
Comment (drwells) [inline ibtk/src/lagrangian/FEDataManager.cpp:2793]: This type of refactoring really should be in another PR but lets work with what we have.

File: include/ibamr/FEMechanicsBase.h [hunk 7]
```diff
@@ -480,6 +530,11 @@ class FEMechanicsBase : public SAMRAI::tbox::Serializable
                                                  unsigned int part,
                                                  const std::string& extension);
 
+    /*!
+     * Cached input databases.
+     */
+    SAMRAI::tbox::Pointer<SAMRAI::tbox::Database> d_fe_data_manager_db, d_fe_projector_db;
```
Comment (drwells) [inline include/ibamr/FEMechanicsBase.h:536]: `FEMechanicsBase` doesn't know anything about `FEDataManager` (nor should it). We should ignore any appearances of "FEDataManager" in the input database to this class (`IBFEMethod` can deal with it itself later). Since the idea of passing an `FEProjector` database is new we can fix the problem without worrying about backwards compatibility. Lets just check if `FEProjector` is a database in the input database to this class - if it is we pass it along, if not we create a new empty database and store that.

File: src/IB/FEMechanicsExplicitIntegrator.cpp [hunk 8]
```diff
@@ -0,0 +1,647 @@
...
+#include "ibtk/FEProjector.h"
+#include "ibtk/libmesh_utilities.h"
+
+#include "libmesh/boundary_info.h"
+#include "libmesh/enum_xdr_mode.h"
+#include "libmesh/equation_systems.h"
+#include "libmesh/quadrature_gauss.h"
+
+#include <IBTK_config.h>
+
```
Comment (drwells) [inline src/IB/FEMechanicsExplicitIntegrator.cpp:31]:
```suggestion
#include <IBTK_config.h>

```
These top-level includes don't exist in the CMake build system.

File: src/IB/FEMechanicsExplicitIntegrator.cpp [hunk 9]
```diff
@@ -0,0 +1,647 @@
...
+// This file is part of IBAMR.
+//
+// IBAMR is free software and is distributed under the 3-clause BSD
+// license. The full text of the license can be found in the file
+// COPYRIGHT at the top level directory of IBAMR.
+//
+// ---------------------------------------------------------------------
+
+/////////////////////////////// INCLUDES /////////////////////////////////////
+
```
Comment (drwells) [inline src/IB/FEMechanicsExplicitIntegrator.cpp:15]:
```suggestion
#include "ibtk/config.h"

```

File: src/IB/FEMechanicsExplicitIntegrator.cpp (hunk 5 above)
Comment (drwells) [reply src/IB/FEMechanicsExplicitIntegrator.cpp:534]: I don't really see a problem with this approach - those two classes don't know anything about each-other and this class doesn't need any IB data so this looks totally fine to me.

File: src/IB/FEMechanicsExplicitIntegrator.cpp [hunk 10]
```diff
@@ -0,0 +1,647 @@
...
+    // We assume that the dofs have the same layout as long as they have the
+    // same fe type and number of variables
+    std::pair<unsigned int, FEType> key(n_vars, fe_type);
+
+    std::unique_ptr<FEData::SystemDofMapCache>& dof_map_cache = d_system_dof_map_cache[part][key];
+    if (dof_map_cache == nullptr)
+    {
+        dof_map_cache.reset(new FEData::SystemDofMapCache(system));
+    }
+    return dof_map_cache.get();
```
Comment (drwells) [inline src/IB/FEMechanicsExplicitIntegrator.cpp:582]: What's up with this function? Is there really no way to share this information (i.e., use the common `FEData` object that contains these things)?

**Review by drwells — CHANGES_REQUESTED** [review]
I think this is mostly done. What is the plan for writing tests?

File: include/ibamr/FEMechanicsBase.h (hunk 7 above)
Comment (boyceg) [reply include/ibamr/FEMechanicsBase.h:536]: I have removed `d_fe_data_manager_db` from `FEMechanicsBase` and moved the checks for those input settings to `IBFEMethod`.

File: src/IB/FEMechanicsExplicitIntegrator.cpp (hunk 8 above)
Comment (boyceg) [reply src/IB/FEMechanicsExplicitIntegrator.cpp:31]: Removed.

File: src/IB/FEMechanicsExplicitIntegrator.cpp (hunk 9 above)
Comment (boyceg) [reply src/IB/FEMechanicsExplicitIntegrator.cpp:15]: Added.

File: src/IB/FEMechanicsExplicitIntegrator.cpp (hunk 10 above)
Comment (boyceg) [reply src/IB/FEMechanicsExplicitIntegrator.cpp:582]: Good question. It does not seem to be used; I will remove it.

**Comment by boyceg** [comment]
@drwells, I have added a test, but I am having trouble getting the output synched up, so currently the test is failing. Can you take a look and tell me what dumb thing I have done?

I think I also need to add the test to the CMakeLists.txt.

**Comment by boyceg** [comment]
@drwells, can you take another look?

**Comment by boyceg** [comment]
This now has simple serial and parallel tests.

**Comment by drwells** [comment]
I'm on it!

File: src/IB/FEMechanicsBase.cpp [hunk 11]
```diff
@@ -526,6 +741,119 @@ FEMechanicsBase::computeStaticPressure(PetscVector<double>& P_vec,
...
+        QBase::build(d_default_quad_type_pressure[part], dim, d_default_quad_order_pressure[part]);
+    fe.attachQuadratureRule(qrule.get());
+    fe.evalQuadraturePoints();
+    fe.evalQuadratureWeights();
+    fe.registerSystem(P_system, P_vars, no_vars);
+    const size_t X_sys_idx = fe.registerInterpolatedSystem(X_system, no_vars, X_vars, &X_vec);
+    const size_t U_sys_idx = fe.registerInterpolatedSystem(U_system, no_vars, U_vars, &U_vec);
+    fe.init();
+
+    const std::vector<libMesh::Point>& q_point = fe.getQuadraturePoints();
```
Comment (drwells) [inline src/IB/FEMechanicsBase.cpp:787]: I don't think we ever use quadrature points so we can skip this.
```suggestion
```

File: src/IB/FEMechanicsBase.cpp [hunk 12]
```diff
@@ -526,6 +741,119 @@ FEMechanicsBase::computeStaticPressure(PetscVector<double>& P_vec,
...
+    const std::vector<double>& JxW = fe.getQuadratureWeights();
+    const std::vector<std::vector<double> >& phi = fe.getPhi(P_fe_type);
+
+    const std::vector<std::vector<std::vector<double> > >& fe_interp_var_data = fe.getVarInterpolation();
```
Comment (drwells) [inline src/IB/FEMechanicsBase.cpp:791]:
```suggestion
```
Same

File: src/IB/FEMechanicsExplicitIntegrator.cpp [hunk 13]
```diff
@@ -0,0 +1,611 @@
...
+
+void
+FEMechanicsExplicitIntegrator::getFromInput(const Pointer<Database>& db, bool /*is_from_restart*/)
+{
+    // Problem parameters.
+    if (db->isDouble("mass_density"))
+        std::fill(d_rhos.begin(), d_rhos.end(), db->getDouble("mass_density"));
+    else if (db->keyExists("mass_density"))
+    {
+        TBOX_ASSERT(db->getArraySize("mass_density") == d_rhos.size());
```
Comment (drwells) [inline src/IB/FEMechanicsExplicitIntegrator.cpp:581]:
```suggestion
        TBOX_ASSERT(static_cast<std::size_t>(db->getArraySize("mass_density")) == d_rhos.size());
```

File: tests/Makefile.in [hunk 14]
```diff
@@ -405,9 +405,15 @@ IBAMR2d_LIBS = ${top_builddir}/lib/libIBAMR2d.a ${top_builddir}/ibtk/lib/libIBTK
...
 SUFFIXES = .f.m4
+<<<<<<< HEAD
 TEST_DIRECTORIES = adv_diff advect CIB coarsen complex_fluids IB IBFE IBTK \
 interpolate level_set multiphase_flow navier_stokes physical_boundary refine \
 spread vc_navier_stokes wave_tank
+=======
+TEST_DIRECTORIES = adv_diff advect CIB complex_fluids fe_mechanics IB \
+IBFE IBTK interpolate level_set multiphase_flow navier_stokes \
+physical_boundary refine spread vc_navier_stokes wave_tank
+>>>>>>> 2ef19e451 (add an example)
```
Comment (drwells) [inline tests/Makefile.in:416]: Just bootstrap and recommit.

File: tests/fe_mechanics/fe_mechanics_ex0.cpp [hunk 15]
```diff
@@ -0,0 +1,548 @@
...
+{
+    PP = bulk_mod * log(FF.det()) * tensor_inverse_transpose(FF, NDIM);
+}
+
+void
+solid_body_force_function(VectorValue<double>& F,
+                          const TensorValue<double>& /*FF*/,
+                          const libMesh::Point& /*x*/,
+                          const libMesh::Point& /*X*/,
+                          Elem* const elem,
```
Comment (drwells) [inline tests/fe_mechanics/fe_mechanics_ex0.cpp:148]:
```suggestion
                          Elem* const /*elem*/,
```

File: tests/fe_mechanics/fe_mechanics_ex0.cpp [hunk 16]
```diff
@@ -0,0 +1,548 @@
...
+                          const vector<const vector<double>*>& var_data,
+                          const vector<const vector<VectorValue<double> >*>& /*grad_var_data*/,
+                          double time,
```
Comment (drwells) [inline tests/fe_mechanics/fe_mechanics_ex0.cpp:151]:
```suggestion
                          double /*time*/,
```

File: tests/fe_mechanics/fe_mechanics_ex0.cpp [hunk 17]
```diff
@@ -0,0 +1,548 @@
...
+        // Parse command line options, set some standard options from the input
+        // file, initialize the restart database (if this is a restarted run),
+        // and enable file logging.
+        Pointer<AppInitializer> app_initializer = new AppInitializer(argc, argv, "IB.log");
+        Pointer<Database> input_db = app_initializer->getInputDatabase();
+
+        // Get various standard options set in the input file.
+        const bool dump_viz_data = app_initializer->dumpVizData();
+        const int viz_dump_interval = app_initializer->getVizDumpInterval();
+        const bool uses_visit = dump_viz_data && app_initializer->getVisItDataWriter();
```
Comment (drwells) [inline tests/fe_mechanics/fe_mechanics_ex0.cpp:227]:
```suggestion
```

File: tests/fe_mechanics/fe_mechanics_ex0.cpp [hunk 18]
```diff
@@ -0,0 +1,548 @@
...
+namespace ModelData
+{
+static BoundaryInfo* lag_bdry_info;
+
+// Tether (penalty) force function for the solid blocks.
+static double kappa = 1.0e6;
+static double damping = 0.0;
+static double traction_force = 6.25e3;
+static double load_time;
+static double PI = 3.141592653589793;
```
Comment (drwells) [inline tests/fe_mechanics/fe_mechanics_ex0.cpp:43]:
```suggestion
```

File: tests/fe_mechanics/fe_mechanics_ex0.cpp [hunk 19]
```diff
@@ -0,0 +1,548 @@
...
+        shear_mod = input_db->getDouble("SHEAR_MOD");
+        bulk_mod = input_db->getDouble("BULK_MOD");
+        damping = input_db->getDouble("ETA");
+        kappa = input_db->getDouble("KAPPA");
+        traction_force = input_db->getDouble("TRACTION_FORCE");
+        load_time = input_db->getDouble("LOAD_TIME");
+
+        use_static_pressure = input_db->getBool("USE_STATIC_PRESSURE");
+        use_dynamic_pressure = input_db->getBool("USE_DYNAMIC_PRESSURE");
+        PressureProjectionType pressure_proj_type;
```
Comment (drwells) [inline tests/fe_mechanics/fe_mechanics_ex0.cpp:288]:
```suggestion
        PressureProjectionType pressure_proj_type = UNKNOWN_PRESSURE_TYPE;
```

File: tests/fe_mechanics/fe_mechanics_ex0.cpp [hunk 20]
```diff
@@ -0,0 +1,548 @@
...
+        {
+            fem_solver->registerStaticPressurePart(pressure_proj_type);
+        }
+        if (use_dynamic_pressure)
+        {
+            fem_solver->registerDynamicPressurePart(pressure_proj_type);
+        }
+        EquationSystems* equation_systems = fem_solver->getEquationSystems();
+        ExplicitSystem& jac_system = equation_systems->add_system<ExplicitSystem>("JacobianDeterminant");
+        unsigned int J_var = jac_system.add_variable("Avg J", CONSTANT, MONOMIAL);
```
Comment (drwells) [inline tests/fe_mechanics/fe_mechanics_ex0.cpp:369]:
```suggestion
jac_system.add_variable("Avg J", CONSTANT, MONOMIAL);
```

**Review by drwells — COMMENTED** [review]
Most of my comments are minor - I'll put up a patch to get this working with CMake in a moment.

We can get rid of `object_name`, `meshes`, `restart_read_dirname`, and `restart_restore_number` in `IBFEMethod::commonConstructor` now that they are set in the base class.

**Comment by boyceg** [comment]
@drwells I think I have implemented the suggested changes.

File: tests/Makefile.in (hunk 14 above)
Comment (boyceg) [reply tests/Makefile.in]: Whoops!

File: src/IB/FEMechanicsBase.cpp [hunk 21]
```diff
@@ -1272,6 +1601,30 @@ FEMechanicsBase::getFromInput(const Pointer<Database>& db, bool /*is_from_restar
...
+    {
+        d_fe_projector_db = db->getDatabase("FEProjector");
+    }
+    else
+    {
+        d_fe_projector_db = new InputDatabase("FEProjector");
+        if (db->keyExists("num_fischer_vectors"))
+        {
+            d_fe_projector_db->putInteger("num_fischer_vectors", db->getInteger("num_fischer_vectors"));
+        }
```
Comment (drwells) [inline src/IB/FEMechanicsBase.cpp:1625]:
```suggestion
```
I'd skip this  - if the user wants a certain number of fischer vectors they should supply an `FEProjector` database.

File: src/IB/IBFEMethod.cpp [hunk 22]
```diff
@@ -1783,14 +1723,26 @@ IBFEMethod::doInitializeFEEquationSystems()
...
+    IntVector<NDIM> min_ghost_width(0);
+    if (!d_primary_eulerian_data_cache) d_primary_eulerian_data_cache = std::make_shared<SAMRAIDataCache>();
+    d_active_eulerian_data_cache = d_primary_eulerian_data_cache;
+    for (unsigned int part = 0; part < d_meshes.size(); ++part)
+    {
+        EquationSystems& equation_systems = *d_equation_systems[part];
         const std::string manager_name = "IBFEMethod FEDataManager::" + std::to_string(part);
         d_fe_data[part] =
             std::make_shared<FEData>(manager_name + "::fe_data", equation_systems, /*register_for_restart*/ true);
+        d_fe_projectors[part] = std::make_shared<FEProjector>(d_fe_data[part], d_fe_projector_db);
```
Comment (drwells) [inline src/IB/IBFEMethod.cpp:1742]:
```suggestion
```
These are set up now in `FEMechanicsBase::doInitializeFEEquationSystems()`.

File: tests/fe_mechanics/fe_mechanics_ex0.cpp (hunk 16 above)
Comment (drwells) [reply tests/fe_mechanics/fe_mechanics_ex0.cpp:150]: I think you may have forgotten to push this change :)

File: tests/fe_mechanics/fe_mechanics_ex0.cpp (hunk 15 above)
Comment (drwells) [reply tests/fe_mechanics/fe_mechanics_ex0.cpp:147]: I think you may have forgotten to push this change :)

File: src/IB/IBFEMethod.cpp [hunk 23]
```diff
@@ -3238,17 +3037,38 @@ IBFEMethod::getFromInput(const Pointer<Database>& db, bool /*is_from_restart*/)
...
+    if (db->keyExists("FEDataManager"))
+    {
+        d_fe_data_manager_db = db->getDatabase("FEDataManager");
+    }
+    else
+    {
+        d_fe_data_manager_db = new InputDatabase("FEDataManager");
+    }
+
+    // FEProjector settings.
```
Comment (drwells) [inline src/IB/IBFEMethod.cpp:3051]: `d_fe_projector_db` is also set up by the base class. We should (until we completely remove it after the release) put the `FEProjector` database inside the `FEDataManager` database should it not be there already.

I recommend we replace this with
```cpp
if (!d_fe_data_manager_db->keyExists("FEProjector"))
{
    d_fe_data_manager_db->putDatabase(d_fe_projector_db);
}
```
and make it clear in the documentation that `FEProjector` should be given as a database inside the one passed to `IBFEMethod`/`FEMechanicsBase` et al and not inside the `FEDataManager` database.

**Review by drwells — CHANGES_REQUESTED** [review]
I found a few minor things to fix. If you agree with the changes I can go ahead and make them and merge myself.

**Comment by boyceg** [comment]
> I found a few minor things to fix. If you agree with the changes I can go ahead and make them and merge myself.

All look good to me!

**Review by drwells — APPROVED** [review]

**Comment by drwells** [comment]
This puts us at 430 tests 🎉
